File: player.go
package main

import (
	"github.com/veandco/go-sdl2/img"
	"github.com/veandco/go-sdl2/sdl"
)

type PlayerState uint8

const (
	Idle PlayerState = iota
	Running
	Jumping
	Falling
	WallSliding
	WallJumping
)

type Player_t struct {
	Rect        *sdl.Rect
	Texture     *sdl.Texture
	TextureArea *sdl.Rect
	State       PlayerState
	Frame       uint8
	Direction   bool
	Inertia     *sdl.Point
}

var playerStateDim map[PlayerState]sdl.Point = map[PlayerState]sdl.Point{
	Idle: {X: 20, Y: 37}, Running: {X: 20, Y: 37}, Jumping: {X: 20, Y: 37}, Falling: {X: 20, Y: 37}, WallSliding: {X: 20, Y: 37}, WallJumping: {X: 20, Y: 37},
}

func CreatePlayer(screen *Screen_t) *Player_t {
	player := &Player_t{
		Rect:    &sdl.Rect{},
		Inertia: &sdl.Point{},
	}
	dim := playerStateDim[player.State]
	player.Rect.W, player.Rect.H = dim.X, dim.Y

	surface, err := img.Load(PlayerSpriteSheet)
	if err != nil {
		panic(err)
	}
	player.Texture, err = screen.Renderer.CreateTextureFromSurface(surface)
	if err != nil {
		panic(err)
	}
	player.TextureArea = &sdl.Rect{
		W: PlayerSpriteWidth,
		H: PlayerSpriteHeight,
	}

	return player
}

func (player *Player_t) Copy(screen *Screen_t) {
	player.setTextureArea()
	flip := sdl.FLIP_NONE
	if player.Direction == Left {
		flip = sdl.FLIP_HORIZONTAL
	}
	dst := &sdl.Rect{
		X: player.Rect.X + player.Rect.W/2 - PlayerSpriteWidth/2,
		Y: player.Rect.Y + player.Rect.H/2 - PlayerSpriteHeight/2,
		W: PlayerSpriteWidth, H: PlayerSpriteHeight,
	}
	if err := screen.Renderer.CopyEx(player.Texture, player.TextureArea, dst, 0, nil, flip); err != nil {
		panic(err)
	}
}

func (player *Player_t) GetRect() *sdl.Rect {
	return player.Rect
}

func (player *Player_t) Destroy() {
	player.Texture.Destroy()
}

func (player *Player_t) SetState(state PlayerState) {
	if player.State != state {
		player.State = state
		player.Frame = 0
	}
}

func (player *Player_t) Update(screen *Screen_t, level *Level_t) {
	state := sdl.GetKeyboardState()

	if player.Inertia.Y >= 0 {
		ground := player.getGround(level)
		if ground != -1 {
			player.Rect.Y = ground - player.Rect.H
			player.Inertia.Y = 0
			player.groundControl(state, level)
			return
		}

		player.SetState(Falling)
	}

	player.airControl(state, level)

	player.Inertia.Y += Gravity
	player.MoveY(player.Inertia.Y/InertiaPerPixel, level)
	if player.Inertia.X > 0 {
		player.Inertia.X -= AirSlow
		player.Inertia.X = Max32(player.Inertia.X, 0)
	} else {
		player.Inertia.X += AirSlow
		player.Inertia.X = Min32(player.Inertia.X, 0)
	}
	player.MoveX(player.Inertia.X/InertiaPerPixel, level)
}

func (player *Player_t) MoveX(delta int32, level *Level_t) {
	projection := &sdl.Rect{
		X: player.Rect.X + delta,
		Y: player.Rect.Y, W: player.Rect.W, H: player.Rect.H,
	}

	player.Rect.X = func() int32 {
		for _, tile := range level.Tiles {
			if projection.HasIntersection(tile.Rect) {
				if delta > 0 {
					return tile.Rect.X - player.Rect.W
				}
				return tile.Rect.X + tile.Rect.W
			}
		}
		union := projection.Union(level.Bounds)
		if !union.Equals(level.Bounds) {
			if delta > 0 {
				return level.Bounds.X + level.Bounds.W - player.Rect.W
			}
			return level.Bounds.X
		}

		return player.Rect.X + delta
	}()
}

func (player *Player_t) MoveY(delta int32, level *Level_t) {
	projection := &sdl.Rect{
		Y: player.Rect.Y + delta,
		X: player.Rect.X, W: player.Rect.W, H: player.Rect.H,
	}

	player.Rect.Y = func() int32 {
		for _, tile := range level.Tiles {
			if projection.HasIntersection(tile.Rect) {
				player.Inertia.Y = 0
				return tile.Rect.Y + tile.Rect.H
			}
		}
		union := projection.Union(level.Bounds)
		if !union.Equals(level.Bounds) {
			player.Inertia.Y = 0
			return level.Bounds.Y
		}

		return player.Rect.Y + delta
	}()
}

func (player *Player_t) Step(direction bool, level *Level_t) {
	player.SetState(Running)

	if direction == Left {
		player.MoveX(-PlayerStep, level)
		player.Inertia.X = -PlayerStep * InertiaPerPixel
	} else {
		player.MoveX(PlayerStep, level)
		player.Inertia.X = PlayerStep * InertiaPerPixel
	}
	player.Direction = direction
}

func (player *Player_t) groundControl(keyState []uint8, level *Level_t) {
	if keyState[sdl.SCANCODE_W] == 1 {
		player.SetState(Jumping)
		player.Inertia.Y = -JumpPower
		return
	}

	var step func(*Player_t, *Level_t)
	moveKeysPressed := 0
	for _, key := range []uint{sdl.SCANCODE_A, sdl.SCANCODE_D} {
		if keyState[key] == 1 {
			step = GroundSteps[sdl.Scancode(key)]
			moveKeysPressed++
		}
	}
	if moveKeysPressed == 1 {
		step(player, level)
	} else {
		player.SetState(Idle)
		player.Inertia.X = 0
	}
}

func (player *Player_t) airControl(keyState []uint8, level *Level_t) {
	if keyState[sdl.SCANCODE_A] == 1 {
		player.Inertia.X -= AirMovePower
		player.Inertia.X = Max32(player.Inertia.X, -PlayerStep*InertiaPerPixel)
	}
	if keyState[sdl.SCANCODE_D] == 1 {
		player.Inertia.X += AirMovePower
		player.Inertia.X = Min32(player.Inertia.X, PlayerStep*InertiaPerPixel)
	}
}

func (player *Player_t) getGround(level *Level_t) int32 {
	groundHitBox := &sdl.Rect{
		X: player.Rect.X,
		Y: player.Rect.Y + player.Rect.H + 1,
		W: player.Rect.W, H: player.Inertia.Y/InertiaPerPixel + 1,
	}
	for _, tile := range level.Tiles {
		if groundHitBox.HasIntersection(tile.Rect) {
			return tile.Rect.Y
		}
	}
	union := groundHitBox.Union(level.Bounds)
	if !union.Equals(level.Bounds) {
		return level.Bounds.Y + level.Bounds.H
	}
	return -1
}

File: cs/client/Player.cs
using System;
using SFML.Graphics;
using SFML.System;

namespace shootingame
{
    public enum PlayerState
    {
        Idle,
        Running,
        Jumping,
        Falling,
        WallSliding,
        WallJumping,
        Swinging
    }

    public class Player
    {
        public Texture Texture;
        public IntRect TextureArea;
        public int ID;
        public string Name;
        public int Score;
        public int Deaths;
        public IntRect Rect;
        public PlayerState State;
        public uint Frame;
        public bool Direction;
        public Vector2i HookPoint;
        public bool Hooked;
        public bool JumpEnabled;
        public Vector2i Inertia;
        public uint GunCoolDown;
        public Vector2i ShotPoint;
        public bool Shot;
        public int ShotCount;
        public bool HasRespawned;

        public void FromLightPlayer(LightPlayer player) {
            ID = player.ID;
            Rect.Left = player.Pos.X;
            Rect.Top = player.Pos.Y;
            State = player.State;
            Frame = player.Frame;
            Direction = player.Direction;
            HookPoint = new Vector2i() {
                X = player.HookPoint.X,
                Y = player.HookPoint.Y
            };
            Hooked = player.Hooked;
        }
        public Player(LightPlayer player) {
            MakeRect();
            FromLightPlayer(player);
        }
        public Player(int id, string name)
        {
            MakeRect();
            
            ID = id;
            Name = name;
            Score = 0;
            Deaths = 0;
            Inertia =  new Vector2i();
            HookPoint = new Vector2i();
            Hooked = false;
            JumpEnabled = true;
            GunCoolDown = 0;
            ShotPoint = new Vector2i();
            Shot = false;
            ShotCount = 0;
            HasRespawned = false;

            Texture = new Texture(Program.ResourceDir + Const.PlayerSpriteSheet);

            TextureArea = new IntRect(0, 0, width: Const.PlayerSpriteWidth, height: Const.PlayerSpriteHeight);
        }

        public void Draw()
        {
            PlayerAnimations.SetTextureArea(this);
            
            Func<int, int> scale = (x) => x * Const.PlayerScalePercent / 100;
            int width = scale(Const.PlayerSpriteWidth), height = scale(Const.PlayerSpriteHeight);
            IntRect dst = Geometry.AdaptRect(new IntRect(
                left: Rect.Left + Rect.Width/2 - width/2,
                top: Rect.Top + Rect.Height/2 - height/2,
                width: width, height: height
            ));

            Screen.GameScene.Draw(Drawing.SpriteOf(Texture, dst, TextureArea));

            if (Hooked) {
                VertexArray line = new VertexArray(PrimitiveType.Lines);
                line.Append(new Vertex(
                    (Vector2f)Geometry.AdaptPoint(GetCOM()),
                    new Color(red: 255, 0, 0)
                ));
                line.Append(new Vertex(
                    (Vector2f)Geometry.AdaptPoint(HookPoint),
                    new Color(red: 255, 0, 0)
                ));
                Screen.GameScene.Draw(line);
            }
        }

        public void Update(Level level)
        {
            ++Frame;

            if (Controls.RightClick) {
                if (!Hooked) {
                    Vector2i hookPoint = HitScan(level, false);
                    if (Geometry.Dist(GetCOM(), hookPoint) <= Const.HookMaxRange) {
                        Hooked = true;
                        HookPoint = hookPoint;
                        Sounds.PlayIrregularFor("hook", Const.HookSoundStep);
                    }
                } else {
                    if (Geometry.Dist(GetCOM(), HookPoint) <= Const.HookMaxRange) {
                        Controls.Swing(this, level);
                        return;
                    }
                    Hooked = false;
                }
            } else {
                Hooked = false;
            }

            Shot = false;
            if (GunCoolDown > 0) --GunCoolDown;
            if (GunCoolDown == 0 && Controls.LeftClick) {
                ShotPoint = HitScan(level, true);
                Shot = true;
                ShotCount += 1;
                GunCoolDown = Const.GunCoolDown;
            }

            if (!JumpEnabled && !Controls.Jump) {
                JumpEnabled = true;
            }

            if (Inertia.Y >= 0) {
                if (OnGround(level)) {
                    Controls.OnGround(this, level);
                    return;
                }
                SetState(PlayerState.Falling);
            }

            MoveX(Inertia.X/Const.InertiaPerPixel, level);
            MoveY(Inertia.Y/Const.InertiaPerPixel, level);
            
            Controls.InAir(this, level);
        }

        public void SetState(PlayerState state)
        {
            if (State != state) {
                State = state;
                Frame = 0;
            }
        }

        public bool OnGround(Level level) {
            IntRect rightUnderFeet = new IntRect(
                left: Rect.Left,
                top: Rect.Top + Rect.Height,
                width: Rect.Width,
                height: 1
            );
            return Controls.Collision(ref rightUnderFeet, level);
        }

        public Vector2i GetCOM()
        {
            return new Vector2i(
                x: Rect.Left + Rect.Width/2,
                y: Rect.Top + Rect.Height/2 
            );
        }
        
        public void MakeRect()
        {
            Func<int, int> scale = (x) => x * Const.PlayerScalePercent / 100;
            Rect = new IntRect(0, 0, width: scale(Const.NormalStateDimX), height: scale(Const.NormalStateDimY));
        }

        private Vector2i HitScan(Level level, bool hitPlayers)
        {
            Vector2i hit = new Vector2i();
            
            Vector2i playerCOM = Geometry.AdaptPoint(GetCOM());
            Vector2i proj = Geometry.PointShade(Game.Bounds, playerCOM, Controls.MousePos.X, Controls.MousePos.Y);
            double minDist = Double.PositiveInfinity;
            
            Action<Vector2i> minimize = (point) => {
                double dist = Geometry.Dist(playerCOM, point);
                if (dist < minDist) {
                    minDist = dist;
                    hit = point;
                }
            };
            minimize(proj);

            foreach (var tile in level.Tiles)
            {
                var rect = Geometry.AdaptRect(tile.Rect);
                if (Geometry.IntersectLine(rect, playerCOM, proj)) {
                    var point = Geometry.HitPoint(rect, playerCOM, proj);
                    minimize(point);
                }
            }
            if (hitPlayers) {
                foreach (var lightPlayer in Game.Players)
                {
                    if (lightPlayer == null || lightPlayer.ID == ID)
                        continue;
                    var player = new Player(lightPlayer);
                    var rect = Geometry.AdaptRect(player.Rect);
                    if (Geometry.IntersectLine(rect, playerCOM, proj)) {
                        var point = Geometry.HitPoint(rect, playerCOM, proj);
                        minimize(point);
                    }
                }
            }
            
            hit.X = (hit.X - Game.Bounds.Left)*level.Bounds.Width/Game.Bounds.Width;
            hit.Y = (hit.Y - Game.Bounds.Top)*level.Bounds.Height/Game.Bounds.Height;
            return hit;
        }

        public void MoveX(int delta, Level level)
        {
            IntRect projection = new IntRect(
                left: Rect.Left + delta,
                top: Rect.Top, width: Rect.Width, height: Rect.Height
            );

            if (Controls.Collision(ref projection, level))
            {
                Inertia.X = 0;
                if (delta > 0) {
                    Rect.Left = projection.Left - Rect.Width;
                } else {
                    Rect.Left = projection.Left + projection.Width;
                }
            }
            else
            {
                Rect.Left = projection.Left;
            }
        }

        public void MoveY(int delta, Level level)
        {
            IntRect projection = new IntRect(
                left: Rect.Left,
                top: Rect.Top + delta,
                width: Rect.Width, height: Rect.Height
            );

            if (Controls.Collision(ref projection, level))
            {
                Inertia.Y = 0;
                if (delta > 0) {
                    Rect.Top = projection.Top - Rect.Height;
                } else {
                    Rect.Top = projection.Top + projection.Height;
                }
            }
            else
            {
                Rect.Top = projection.Top;
            }
        }
    }
}
